lib/membrane_rtc_engine/endpoints/rtsp/connection_manager.ex

```diff
@@ -0,0 +1,317 @@
+    end
+
+    defp start_rtsp_session(%ConnectionStatus{rtsp_session: nil, stream_uri: stream_uri}) do
+      case RTSP.start(stream_uri) do
+        {:ok, session} ->
+          Process.monitor(session)
+          session
+
+        {:error, error} ->
+          Membrane.Logger.warn(
+            "ConnectionManager: Starting RTSP session failed - #{inspect(error)}"
+          )
+
+          nil
+      end
+    end
+
+    defp start_rtsp_session(%ConnectionStatus{rtsp_session: rtsp_session}) do
+      rtsp_session
+    end
+
+    defp get_rtsp_description(%ConnectionStatus{rtsp_session: rtsp_session} = connection_status) do
+      Membrane.Logger.debug("ConnectionManager: Setting up RTSP description")
+
+      case RTSP.describe(rtsp_session) do
+        {:ok, %{status: 200, body: %{media: sdp_media}}} ->
+          attributes = get_video_attributes(sdp_media)
+
+          [fmtp: attributes["fmtp"], control: attributes["control"], rtpmap: attributes["rtpmap"]]
+          |> then(
+            &{:ok,
+             Map.update(connection_status, :endpoint_options, [], fn endpoint_options ->
+               Keyword.merge(endpoint_options, &1)
+             end)}
+          )
+
+        _result ->
+          {:error, :getting_rtsp_description_failed}
+      end
+    end
+
+    defp setup_rtsp_connection(
+           %ConnectionStatus{
+             rtsp_session: rtsp_session,
+             endpoint_options: endpoint_options
+           } = connection_status
+         ) do
+      Membrane.Logger.debug("ConnectionManager: Setting up RTSP connection")
+
+      case RTSP.setup(rtsp_session, "/#{endpoint_options[:control]}", [
+             {"Transport", "RTP/AVP;unicast;client_port=#{endpoint_options[:port]}"}
+           ]) do
+        {:ok, %{status: 200, headers: headers}} ->
+          # Parse server port to be able to combat NAT later
+          server_port = parse_server_port(headers)
+
+          {:ok,
+           %{
+             connection_status
+             | endpoint_options: Keyword.merge(endpoint_options, server_port: server_port)
+           }}
+
+        result ->
+          Membrane.Logger.debug(
+            "ConnectionManager: Setting up RTSP connection failed: #{inspect(result)}"
+          )
+
+          {:error, :setting_up_sdp_connection_failed}
+      end
+    end
+
+    defp play(%ConnectionStatus{rtsp_session: rtsp_session, endpoint: _endpoint}) do
+      Membrane.Logger.debug("ConnectionManager: Setting RTSP on play mode")
+
+      case RTSP.play(rtsp_session) do
+        {:ok, %{status: 200}} ->
+          :ok
+
+        _result ->
+          {:error, :play_rtsp_failed}
+      end
+    end
+
+    defp start_keep_alive(%ConnectionStatus{rtsp_session: rtsp_session} = connection_status) do
+      Membrane.Logger.debug("ConnectionManager: Starting Keep alive process")
+
+      {:ok, keep_alive} =
+        Task.start(fn ->
+          rtsp_keep_alive(rtsp_session)
+        end)
+
+      Process.monitor(keep_alive)
```

Comment:
This was a leftover from the original ConnectionManager used in rtsp_to_hls demo -- I didn't touch the parts that worked, but ofc can take a shot at refactoring

Reply:
I think I see why this was done as a separate process. It's problematic because it executes `RTSP.get_parameter`, which is a call -- so, worst case scenario (connection loss), ConnectionManager is blocked for 5 seconds, after which it crashes

```elixir
    14:28:27.593 [error] GenServer #PID<0.2841.0> terminating
     ** (stop) exited in: GenServer.call(#PID<0.2842.0>, {:execute, %Membrane.RTSP.Request{method: "GET_PARAMETER", path: nil, headers: [], body: ""}}, 5000)
         ** (EXIT) time out
         (elixir 1.14.2) lib/gen_server.ex:1038: GenServer.call/3

[...]
```

The way I see it, if we don't use a separate process, we either have to change the implementation of `membrane_rtsp`, rescue or let it crash. If we rescue and handle, ConnectionManager will still block for 5 secs, which right now isn't a big issue, but could become one in the future. If we let it crash, we'll be unable to directly inform Engine's parent about the reason -- no way to send `:disconnected` message. Of course, further changes could be made to handle ConnectionManager crashing... etc etc.

I'm not a big fan of either of these solutions, tbh. I would propose instead to either leave the task as-is, or create another GenServer module which will be responsible solely for the keep-alives, and thus free to crash at will. Is that an option you'd be alright with?